snuba/query/processors/mapping_optimizer.py

```diff
@@ -159,20 +185,110 @@ def __replace_with_hash(self, condition: Expression) -> Expression:
             ),
         )
 
+    def _get_condition_without_redundant_checks(
+        self, condition: Expression, query: Query
+    ) -> Expression:
+        """Optimizes the case where the query condition contains the following:
+
+        valueOf('my_tag') != '' AND valueOf('my_tag') == "something"
```

Comment:
I think this function is missing one use case that we actually instructed the product team to prefer:
`has(tags.key, 'blabla')` and `NOT has(tags.key, 'blabla')`.
We are supposed to see more of these and fewer `valeOf(tags. values, indexOf(tags.key, 'blabla') != ''`.
So we should support both cases now.

Reply:
There is also a different processor (EmptyTagConditionProcessor) that will convert `valueOf(tags. values, indexOf(tags.key, 'blabla') != ''` into `has(tags.key, 'blabla')`. Right now that processor happens after this one, but better to be safe than sorry.